maintainers/README: add guidelines for committers

Scrumplex opened the pull request:
This PR adds some guidelines for (new) committers. Notably this tries to describe what a breaking change is.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Scrumplex committed:
maintainers/README: add guidelines for committers

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex commented:
Note that this PR is very much a draft and also doesn't reflect the final structure.

Mindavi commented on maintainers/README.md:
```diff
@@ -172,3 +172,34 @@ to the team without an approval by at least one existing member.
+branch. If a commit breaks evaluation in `master` it could make it into new PR
+branches and cause further breakage.
+
+One approach to avoid merging such problematic changes is to wait for successful Borg evaluation.
```
Ofborg

Mindavi commented on maintainers/README.md:
```diff
@@ -172,3 +172,34 @@ to the team without an approval by at least one existing member.
+- `unmaintained-software: drop` - If this PR removes a leaf package or the
+  removal doesn't otherwise break other packages, it can be merged.
+- `cool-tool: rename from fancy-tool` - As long as this PR replaces all
+  references to the old attribute name with the new name, it can be merged.
```
And maybe adds an alias

jian-lin commented on maintainers/README.md:
```diff
@@ -172,3 +172,34 @@ to the team without an approval by at least one existing member.
+These are some example changes and if they are considered a breaking change
+during a freeze period:
+
+- `foo: 1.2.3 -> 1.2.4` - Assuming this package follows semantic versioning and
```
What about adding another example about bumping leaf packages with breaking changes.  I used to think they are not allowed during the freeze period.  However, according to the intention of the freeze period, they presumably are allowed.

Frontear commented on maintainers/README.md:
```diff
@@ -172,3 +172,34 @@ to the team without an approval by at least one existing member.
+branch. If a commit breaks evaluation in `master` it could make it into new PR
+branches and cause further breakage.
+
+One approach to avoid merging such problematic changes is to wait for successful Borg evaluation.
```
Maybe we should directly encourage the usage of `nixpkgs-review` here too? Ofborg can take a long while to get around to an evaluation, even more so if the derivation is for darwin. Encouraging a self-test would be great, plus it lets contributers know that there are other things they can do besides wait for Ofborg.

Scrumplex replied:
So my main reason to wait for Borg was to avoid commits like https://github.com/NixOS/nixpkgs/commit/60900303b5ebdbf43ede288b704b655b33d5e4a5 breaking Darwin eval on master. Thoughg interestingly, the eval failure wasn't even caught by Ofborg before I merged that PR.

Frontear replied:
Oh for sure! Ofborg eval is very helpful especially because it will evaluate all the systems, no denying its usefulness. I think it would just be nice to also add a little note about `nixpkgs-review`, mainly because as it is now, its briefly mentioned in the PR template and is surrounded by all the other checkboxes that it can be easily missed. Having it documented standalone and encouraged for every new PR would be a great practice for new contributers to pick up

Frontear commented:
Change looks good! I wonder if we could use this PR to introduce some of the changes @infinisil had mentioned over at https://github.com/NixOS/nixpkgs/issues/347586?

Aleksanaa commented on maintainers/README.md:
```diff
@@ -172,3 +172,39 @@ to the team without an approval by at least one existing member.
+
+# Guidelines for Committers
+
+When merging changes, care must be taken to reduce impact to the `master`
```
```suggestion
When merging pull requests, care must be taken to reduce impact to the `master`
```

Aleksanaa commented on maintainers/README.md:
```diff
@@ -172,3 +172,39 @@ to the team without an approval by at least one existing member.
+# Guidelines for Committers
+
+When merging changes, care must be taken to reduce impact to the `master`
+branch. If a commit breaks evaluation in `master` it could make it into new PR
+branches and cause further breakage.
```
```suggestion
branch. If a commit breaks evaluation, it will affect Ofborg evaluation results
in other pull requests and block hydra CI, thus introducing chaos to our workflow.
```

Aleksanaa commented on maintainers/README.md:
```diff
@@ -172,3 +172,39 @@ to the team without an approval by at least one existing member.
+
+## Breaking changes
+
+In general breaking changes to `master` and `staging` branches are permitted.
+Though restrictions might apply towards the end of a NixOS release cycle. This
+is to avoid large-scale breakages shortly before or during a Zero Hydra
+Failures (ZHF) campaign. These restrictions also intend to decrease the
+likelihood of a delayed NixOS release.
```
```suggestion
In general breaking changes to `master` and `staging` branches are permitted.
Though restrictions might apply towards the end of a NixOS release cycle, due to
our feature freeze mechanism.
This is to avoid large-scale breakages shortly before and during a Zero Hydra
Failures (ZHF) campaign. These restrictions also intend to decrease the
likelihood of a delayed NixOS release. The feature freeze period is documented
in the announcement of each release schedule.
```

Aleksanaa commented on maintainers/README.md:
```diff
@@ -172,3 +172,39 @@ to the team without an approval by at least one existing member.
+
+- `foo: 1.2.3 -> 1.2.4` - Assuming this package follows semantic versioning and
+  none of its dependent packages fail to build because of this change, it can be
+  safely merged.
```
```suggestion
  safely merged. If it can be confirmed that there is no major change in
  its functionality or API, but only adding new features or fixing bugs, it
  can also be merged.
```

Aleksanaa commented on maintainers/README.md:
```diff
@@ -172,3 +172,39 @@ to the team without an approval by at least one existing member.
+  is over. Alternatively, if this PR is for a popular package and doesn't cause
+  many rebuilds, it should also be delayed to reduce risk of breakage. If a PR
+  includes important changes, such as security fixes, it should be brought up to
+  release managers.
```
```suggestion
  release managers.
- `nixos/transmission: refactor` - If this PR adjusts the type, default value or effect of
   options in the NixOS module, so that users must rewrite their configuration to keep
   the current behavior unchanged, it should not be merged, as we don't have enough
   time to collect user feedback and avoid possible breakage. However, it should be
   accepted if the previous behavior associated with the modifications has completely
   broken.
```

Scrumplex commented:
I will be adding future changes as fixup commits. Once we are confident, I can run `git rebase --autosquash` to merge them all into a single commit

Frontear commented on maintainers/README.md:
```diff
@@ -172,3 +172,49 @@ to the team without an approval by at least one existing member.
+workflow.
+
+One approach to avoid merging such problematic changes is to wait for
+successful Ofborg evaluation. Additionally, using tools like `nixpkgs-review`
```
Just thought about it, could/should we change \`nixpkgs-review\` to \[nixpkgs-review\]\(https:\/\/github.com/Mic92/nixpkgs-review)?

Frontear commented on maintainers/README.md:
```diff
@@ -172,3 +172,49 @@ to the team without an approval by at least one existing member.
+In general breaking changes to `master` and `staging` branches are permitted.
+Though restrictions might apply towards the end of a NixOS release cycle, due
+to our feature freeze mechanism. This is to avoid large-scale breakages shortly
+before and during a Zero Hydra Failures (ZHF) campaign. These restrictions also
```
should we explain what "ZHF" is, or is that too out of scope? Maybe instead we could link to somewhere that mentions ZHF?

Scrumplex replied:
I think it's usually explained in the release schedule/ZHF issue and as that is a moving target I think it's okay to not link to anything here

Mindavi commented on maintainers/README.md:
```diff
@@ -172,3 +172,49 @@ to the team without an approval by at least one existing member.
+intend to decrease the likelihood of a delayed NixOS release. The feature
+freeze period is documented in the announcement of each release schedule.
+
+These are some example changes and if they are considered a breaking change
```
This could be a nested paragraph, then it's clearer it's separate from the other guidelines and only counts for a limited time (around releases).

Mindavi commented on maintainers/README.md:
```diff
@@ -172,3 +172,49 @@ to the team without an approval by at least one existing member.
+
+## Breaking changes
+
+In general breaking changes to `master` and `staging` branches are permitted.
```
Mentioning notable breaking charges in the release notes is nice though.

Mindavi reviewed (approved):
Good as a first version, I'd say. But let's try and keep it living documentation give it some love from time to time.

Scrumplex commented:
Rebased, squashed, and resolved conflict with new nixpkgs-review-bot section. This should be good to merge

Scrumplex commented:
Rebased, squashed, and resolved conflict with new nixpkgs-review-bot section. This should be good to be merged. If there are no objections I will merge this in a few hours
